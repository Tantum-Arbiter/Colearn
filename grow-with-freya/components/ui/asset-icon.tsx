import React from 'react';
import { View, StyleSheet, Image } from 'react-native';

interface AssetIconProps {
  source: any;
  size?: number;
  color?: string;
  style?: any;
}

export function AssetIcon({ source, size = 24, color, style }: AssetIconProps) {
  return (
    <View style={[styles.container, style]}>
      <Image 
        source={source} 
        style={[
          { width: size, height: size },
          color && { tintColor: color }
        ]} 
        resizeMode="contain"
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
  },
});


