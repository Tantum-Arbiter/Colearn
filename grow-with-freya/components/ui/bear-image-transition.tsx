import React, { useEffect, useState } from 'react';
import { View, StyleSheet } from 'react-native';
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  withDelay,
  Easing,
} from 'react-native-reanimated';

interface BearImageTransitionProps {
  children: React.ReactNode;
  isVisible: boolean;
  preloadDelay?: number;
  fadeDuration?: number;
}

/**
 * BearImageTransition specifically handles smooth transitions for the bear image
 * Provides a fade-in effect that masks any loading delays
 */
export const BearImageTransition: React.FC<BearImageTransitionProps> = ({
  children,
  isVisible,
  preloadDelay = 200,
  fadeDuration = 400,
}) => {
  const opacity = useSharedValue(0);
  const [shouldRender, setShouldRender] = useState(false);

  useEffect(() => {
    if (isVisible) {
      // Give bear image time to preload
      setTimeout(() => {
        setShouldRender(true);
        
        // Start fade-in animation
        opacity.value = withTiming(1, {
          duration: fadeDuration,
          easing: Easing.out(Easing.ease),
        });
      }, preloadDelay);
    } else {
      // Fade out quickly
      opacity.value = withTiming(0, {
        duration: 200,
        easing: Easing.in(Easing.ease),
      });
      
      // Hide after fade out
      setTimeout(() => setShouldRender(false), 200);
    }
  }, [isVisible, preloadDelay, fadeDuration]);

  const animatedStyle = useAnimatedStyle(() => {
    return {
      opacity: opacity.value,
    };
  });

  if (!shouldRender) {
    return null;
  }

  return (
    <Animated.View style={[styles.container, animatedStyle]}>
      {children}
    </Animated.View>
  );
};

/**
 * MainMenuTransition provides smooth entrance for the main menu
 * Specifically designed to hide bear image loading issues
 */
interface MainMenuTransitionProps {
  children: React.ReactNode;
  isActive: boolean;
}

export const MainMenuTransition: React.FC<MainMenuTransitionProps> = ({
  children,
  isActive,
}) => {
  const containerOpacity = useSharedValue(0);
  const bearOpacity = useSharedValue(0);
  const backgroundOpacity = useSharedValue(0);

  useEffect(() => {
    if (isActive) {
      // Staggered animation for smooth appearance
      
      // 1. Background appears first (fast)
      backgroundOpacity.value = withTiming(1, {
        duration: 200,
        easing: Easing.out(Easing.ease),
      });

      // 2. Main container appears
      containerOpacity.value = withDelay(100, withTiming(1, {
        duration: 300,
        easing: Easing.out(Easing.ease),
      }));

      // 3. Bear image appears last (after preload time)
      bearOpacity.value = withDelay(300, withTiming(1, {
        duration: 400,
        easing: Easing.out(Easing.ease),
      }));
    } else {
      // Quick fade out
      const quickFade = {
        duration: 200,
        easing: Easing.in(Easing.ease),
      };
      
      containerOpacity.value = withTiming(0, quickFade);
      bearOpacity.value = withTiming(0, quickFade);
      backgroundOpacity.value = withTiming(0, quickFade);
    }
  }, [isActive]);

  const containerStyle = useAnimatedStyle(() => ({
    opacity: containerOpacity.value,
  }));

  const bearStyle = useAnimatedStyle(() => ({
    opacity: bearOpacity.value,
  }));

  const backgroundStyle = useAnimatedStyle(() => ({
    opacity: backgroundOpacity.value,
  }));

  return (
    <View style={styles.mainMenuContainer}>
      {/* Background layer */}
      <Animated.View style={[styles.backgroundLayer, backgroundStyle]}>
        {/* Background elements like clouds and stars */}
      </Animated.View>

      {/* Main content layer */}
      <Animated.View style={[styles.contentLayer, containerStyle]}>
        {children}
      </Animated.View>

      {/* Bear image layer - appears last */}
      <Animated.View style={[styles.bearLayer, bearStyle]} pointerEvents="none">
        {/* Bear image will be rendered here */}
      </Animated.View>
    </View>
  );
};

/**
 * Simple fade transition for page changes
 */
interface FadeTransitionProps {
  children: React.ReactNode;
  isVisible: boolean;
  duration?: number;
  delay?: number;
}

export const FadeTransition: React.FC<FadeTransitionProps> = ({
  children,
  isVisible,
  duration = 300,
  delay = 0,
}) => {
  const opacity = useSharedValue(isVisible ? 1 : 0);

  useEffect(() => {
    const animationConfig = {
      duration,
      easing: Easing.inOut(Easing.ease),
    };

    if (delay > 0) {
      opacity.value = withDelay(delay, withTiming(isVisible ? 1 : 0, animationConfig));
    } else {
      opacity.value = withTiming(isVisible ? 1 : 0, animationConfig);
    }
  }, [isVisible, duration, delay]);

  const animatedStyle = useAnimatedStyle(() => ({
    opacity: opacity.value,
  }));

  return (
    <Animated.View style={[styles.container, animatedStyle]}>
      {children}
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  mainMenuContainer: {
    flex: 1,
    position: 'relative',
  },
  backgroundLayer: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    zIndex: 1,
  },
  contentLayer: {
    flex: 1,
    zIndex: 2,
  },
  bearLayer: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
    height: '15%',
    zIndex: 3,
  },
});
