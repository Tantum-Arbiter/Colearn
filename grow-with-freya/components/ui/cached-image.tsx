import React, { useState, useEffect } from 'react';
import { Image, ImageProps, ImageSourcePropType, View, ActivityIndicator } from 'react-native';

interface CachedImageProps extends Omit<ImageProps, 'source'> {
  source: ImageSourcePropType;
  fallbackSource?: ImageSourcePropType;
  showLoadingIndicator?: boolean;
  loadingIndicatorColor?: string;
  preload?: boolean;
  instantDisplay?: boolean;
}

/**
 * CachedImage component that aggressively caches images for instant display
 * Optimized for performance-critical images like the bear image
 */
export const CachedImage: React.FC<CachedImageProps> = ({
  source,
  fallbackSource,
  showLoadingIndicator = false,
  loadingIndicatorColor = '#666',
  preload = true,
  instantDisplay = true,
  style,
  ...props
}) => {
  const [isLoaded, setIsLoaded] = useState(false);
  const [hasError, setHasError] = useState(false);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    if (preload && typeof source === 'object' && 'uri' in source && source.uri) {
      setIsLoading(true);
      Image.prefetch(source.uri)
        .then(() => {
          setIsLoaded(true);
          setIsLoading(false);
        })
        .catch(() => {
          setHasError(true);
          setIsLoading(false);
        });
    } else {
      // For bundled assets (require), they're immediately available
      setIsLoaded(true);
    }
  }, [source, preload]);

  const handleLoad = () => {
    setIsLoaded(true);
    setIsLoading(false);
    setHasError(false);
  };

  const handleError = () => {
    setHasError(true);
    setIsLoading(false);
  };

  const imageSource = hasError && fallbackSource ? fallbackSource : source;

  if (showLoadingIndicator && isLoading && !isLoaded) {
    return (
      <View style={[style, { justifyContent: 'center', alignItems: 'center' }]}>
        <ActivityIndicator color={loadingIndicatorColor} />
      </View>
    );
  }

  return (
    <Image
      {...props}
      source={imageSource}
      style={style}
      onLoad={handleLoad}
      onError={handleError}
      fadeDuration={instantDisplay ? 0 : props.fadeDuration || 200}
      // Optimize for performance
      resizeMethod="resize"
      loadingIndicatorSource={undefined}
      progressiveRenderingEnabled={false}
    />
  );
};

/**
 * Specialized component for the bear image with optimal caching
 */
export const CachedBearImage: React.FC<{ style?: any }> = ({ style }) => {
  return (
    <CachedImage
      source={require('@/assets/images/ui-elements/bear-bottom-screen.png')}
      style={[
        {
          width: 286,
          height: 286,
          opacity: 0.8,
        },
        style,
      ]}
      resizeMode="contain"
      preload={true}
      instantDisplay={true}
    />
  );
};
