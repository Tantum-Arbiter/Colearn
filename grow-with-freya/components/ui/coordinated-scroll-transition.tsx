import React, { useEffect, useState } from 'react';
import { View, StyleSheet, Dimensions } from 'react-native';
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  Easing,
  runOnJS,
} from 'react-native-reanimated';

// Dynamic screen dimensions
const getScreenDimensions = () => {
  const { width, height } = Dimensions.get('window');
  return { width, height };
};

interface CoordinatedScrollTransitionProps {
  currentPage: 'main' | 'stories';
  mainMenuComponent: React.ReactNode;
  storiesComponent: React.ReactNode;
  transitionDuration?: number;
  onTransitionComplete?: (page: 'main' | 'stories') => void;
}

/**
 * CoordinatedScrollTransition creates a seamless scroll effect between pages
 * - When going to stories: main menu scrolls down, stories page scrolls down into view
 * - When going back: stories page scrolls down out of view, main menu scrolls up into view
 */
export const CoordinatedScrollTransition: React.FC<CoordinatedScrollTransitionProps> = ({
  currentPage,
  mainMenuComponent,
  storiesComponent,
  transitionDuration = 600,
  onTransitionComplete,
}) => {
  const [isTransitioning, setIsTransitioning] = useState(false);
  const [previousPage, setPreviousPage] = useState<'main' | 'stories'>(currentPage);

  // Get screen dimensions
  const { height: screenHeight } = getScreenDimensions();

  // Animation values for both pages
  const mainMenuTranslateY = useSharedValue(currentPage === 'main' ? 0 : -screenHeight);
  const storiesTranslateY = useSharedValue(currentPage === 'stories' ? 0 : screenHeight);

  useEffect(() => {
    if (previousPage !== currentPage) {
      performScrollTransition(previousPage, currentPage);
      setPreviousPage(currentPage);
    }
  }, [currentPage]);

  const performScrollTransition = (fromPage: 'main' | 'stories', toPage: 'main' | 'stories') => {
    setIsTransitioning(true);

    const animationConfig = {
      duration: transitionDuration,
      easing: Easing.out(Easing.cubic),
    };

    if (toPage === 'stories') {
      // Going to stories: both pages scroll down
      // Main menu scrolls down and out of view
      mainMenuTranslateY.value = withTiming(-screenHeight, animationConfig);

      // Stories page scrolls down into view (from below screen)
      storiesTranslateY.value = withTiming(0, animationConfig, (finished) => {
        if (finished) {
          runOnJS(() => {
            setIsTransitioning(false);
            onTransitionComplete?.(toPage);
          })();
        }
      });
    } else {
      // Going back to main: stories scrolls down out, main scrolls up in
      // Stories page scrolls down and out of view
      storiesTranslateY.value = withTiming(screenHeight, animationConfig);
      
      // Main menu scrolls up into view (from above screen)
      mainMenuTranslateY.value = withTiming(0, animationConfig, (finished) => {
        if (finished) {
          runOnJS(() => {
            setIsTransitioning(false);
            onTransitionComplete?.(toPage);
          })();
        }
      });
    }
  };

  const mainMenuAnimatedStyle = useAnimatedStyle(() => ({
    transform: [{ translateY: mainMenuTranslateY.value }],
  }));

  const storiesAnimatedStyle = useAnimatedStyle(() => ({
    transform: [{ translateY: storiesTranslateY.value }],
  }));

  return (
    <View style={styles.container}>
      {/* Main Menu Page */}
      <Animated.View 
        style={[styles.page, mainMenuAnimatedStyle]}
        pointerEvents={currentPage === 'main' && !isTransitioning ? 'auto' : 'none'}
      >
        {mainMenuComponent}
      </Animated.View>

      {/* Stories Page */}
      <Animated.View 
        style={[styles.page, storiesAnimatedStyle]}
        pointerEvents={currentPage === 'stories' && !isTransitioning ? 'auto' : 'none'}
      >
        {storiesComponent}
      </Animated.View>
    </View>
  );
};

/**
 * ScrollTransitionContainer provides a simpler interface for the coordinated scroll
 */
interface ScrollTransitionContainerProps {
  children: React.ReactNode;
  pageType: 'main' | 'stories';
  isActive: boolean;
  transitionDirection?: 'up' | 'down';
  duration?: number;
}

export const ScrollTransitionContainer: React.FC<ScrollTransitionContainerProps> = ({
  children,
  pageType,
  isActive,
  transitionDirection = 'down',
  duration = 600,
}) => {
  const translateY = useSharedValue(0);

  useEffect(() => {
    const animationConfig = {
      duration,
      easing: Easing.out(Easing.cubic),
    };

    if (isActive) {
      // Page is becoming active - scroll into view
      translateY.value = withTiming(0, animationConfig);
    } else {
      // Page is becoming inactive - scroll out of view
      const { height: screenHeight } = getScreenDimensions();
      const exitPosition = transitionDirection === 'up' ? -screenHeight : screenHeight;
      translateY.value = withTiming(exitPosition, animationConfig);
    }
  }, [isActive, transitionDirection, duration]);

  const animatedStyle = useAnimatedStyle(() => ({
    transform: [{ translateY: translateY.value }],
  }));

  return (
    <Animated.View 
      style={[styles.page, animatedStyle]}
      pointerEvents={isActive ? 'auto' : 'none'}
    >
      {children}
    </Animated.View>
  );
};

/**
 * VerticalPageTransition provides vertical scroll transitions between pages
 */
interface VerticalPageTransitionProps {
  currentPage: 'main' | 'stories';
  pages: {
    main: React.ReactNode;
    stories: React.ReactNode;
  };
  duration?: number;
}

/**
 * MultiPageTransition provides vertical scroll transitions between multiple pages
 * Extends the stories transition pattern to all navigation destinations
 */
type PageKey = 'main' | 'stories' | 'sensory' | 'emotions' | 'bedtime' | 'screen_time' | 'settings';

interface MultiPageTransitionProps {
  currentPage: PageKey;
  pages: Record<PageKey, React.ReactNode>;
  duration?: number;
}

export const VerticalPageTransition: React.FC<VerticalPageTransitionProps> = ({
  currentPage,
  pages,
  duration = 600,
}) => {
  const { height: screenHeight } = getScreenDimensions();
  const mainTranslateY = useSharedValue(currentPage === 'main' ? 0 : -screenHeight);
  const storiesTranslateY = useSharedValue(currentPage === 'stories' ? 0 : screenHeight);

  useEffect(() => {
    const animationConfig = {
      duration,
      easing: Easing.out(Easing.cubic),
    };

    if (currentPage === 'main') {
      // Show main menu, hide stories
      mainTranslateY.value = withTiming(0, animationConfig);
      storiesTranslateY.value = withTiming(screenHeight, animationConfig);
    } else {
      // Show stories, hide main menu
      mainTranslateY.value = withTiming(-screenHeight, animationConfig);
      storiesTranslateY.value = withTiming(0, animationConfig);
    }
  }, [currentPage, duration]);

  const mainAnimatedStyle = useAnimatedStyle(() => ({
    transform: [{ translateY: mainTranslateY.value }],
  }));

  const storiesAnimatedStyle = useAnimatedStyle(() => ({
    transform: [{ translateY: storiesTranslateY.value }],
  }));

  return (
    <View style={styles.container}>
      {/* Always render main page - positioned off-screen when not active */}
      <Animated.View
        style={[styles.page, mainAnimatedStyle]}
        pointerEvents={currentPage === 'main' ? 'auto' : 'none'}
      >
        {pages.main}
      </Animated.View>

      {/* Always render stories page - positioned off-screen when not active */}
      <Animated.View
        style={[styles.page, storiesAnimatedStyle]}
        pointerEvents={currentPage === 'stories' ? 'auto' : 'none'}
      >
        {pages.stories}
      </Animated.View>
    </View>
  );
};

export const MultiPageTransition: React.FC<MultiPageTransitionProps> = ({
  currentPage,
  pages,
  duration = 800,
}) => {
  // Create shared values for all possible pages
  const pageAnimations = {
    main: useSharedValue(currentPage === 'main' ? 0 : -SCREEN_HEIGHT),
    stories: useSharedValue(currentPage === 'stories' ? 0 : SCREEN_HEIGHT),
    sensory: useSharedValue(currentPage === 'sensory' ? 0 : SCREEN_HEIGHT),
    emotions: useSharedValue(currentPage === 'emotions' ? 0 : SCREEN_HEIGHT),
    bedtime: useSharedValue(currentPage === 'bedtime' ? 0 : SCREEN_HEIGHT),
    screen_time: useSharedValue(currentPage === 'screen_time' ? 0 : SCREEN_HEIGHT),
    settings: useSharedValue(currentPage === 'settings' ? 0 : SCREEN_HEIGHT),
  };

  useEffect(() => {
    const animationConfig = {
      duration,
      easing: Easing.out(Easing.cubic),
    };

    // Animate all pages to their target positions
    Object.keys(pageAnimations).forEach((pageKey) => {
      const key = pageKey as PageKey;
      if (key === currentPage) {
        // Current page slides into view (position 0)
        pageAnimations[key].value = withTiming(0, animationConfig);
      } else if (key === 'main') {
        // Main menu slides up when not active
        pageAnimations[key].value = withTiming(-SCREEN_HEIGHT, animationConfig);
      } else {
        // Other pages slide down when not active
        pageAnimations[key].value = withTiming(SCREEN_HEIGHT, animationConfig);
      }
    });
  }, [currentPage, duration]);

  // Create animated styles for each page
  const createAnimatedStyle = (pageKey: PageKey) => {
    return useAnimatedStyle(() => ({
      transform: [{ translateY: pageAnimations[pageKey].value }],
    }));
  };

  return (
    <View style={styles.container}>
      {/* Render all pages that are provided */}
      {Object.entries(pages).map(([pageKey, pageComponent]) => {
        const key = pageKey as PageKey;
        return (
          <Animated.View
            key={key}
            style={[styles.page, createAnimatedStyle(key)]}
            pointerEvents={currentPage === key ? 'auto' : 'none'}
          >
            {pageComponent}
          </Animated.View>
        );
      })}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    position: 'relative',
    overflow: 'hidden',
    backgroundColor: '#f8f9ff', // Light background to prevent black flickers
  },
  page: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: getScreenDimensions().width,
    height: getScreenDimensions().height,
    backgroundColor: '#f8f9ff', // Match container background
  },
});
