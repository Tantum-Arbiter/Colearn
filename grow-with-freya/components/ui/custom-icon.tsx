import React from 'react';
import { View, StyleSheet, Image } from 'react-native';
import { SvgXml } from 'react-native-svg';

interface CustomIconProps {
  name: string;
  size?: number;
  color?: string;
  style?: any;
  type?: 'svg' | 'png' | 'auto';
}

export function CustomIcon({ name, size = 24, color, style, type = 'auto' }: CustomIconProps) {

  const getIconPath = (iconName: string) => {
    const iconPaths: { [key: string]: any } = {

      'stories': require('@/assets/images/menu-icons/stories.svg'),
      'storybook': require('@/assets/images/menu-icons/storybook.svg'),
      'sparkle-hand': require('@/assets/images/menu-icons/sparkle-hand.svg'),
      'smiley-face': require('@/assets/images/menu-icons/smiley-face.svg'),
      'music-note': require('@/assets/images/menu-icons/music-note.svg'),
      'clock': require('@/assets/images/menu-icons/clock.svg'),
      'gear': require('@/assets/images/menu-icons/gear.svg'),
      'home': require('@/assets/images/menu-icons/home.svg'),
      'activities': require('@/assets/images/menu-icons/activities.svg'),
      'profile': require('@/assets/images/menu-icons/profile.svg'),

      // PNG Menu icons - Pre-loaded for performance
      'stories-icon': require('@/assets/images/menu-icons/stories-icon.png'),
      'sensory-icon': require('@/assets/images/menu-icons/sensory-icon.png'),
      'emotions-icon': require('@/assets/images/menu-icons/emotions-icon.png'),
      'bedtime-icon': require('@/assets/images/menu-icons/bedtime-icon.png'),
      'screentime-icon': require('@/assets/images/menu-icons/screentime-icon.png'),

      // Characters - Only existing files
      'blue-monster-mascot': require('@/assets/images/characters/blue-monster-mascot.svg'),
    };

    return iconPaths[iconName];
  };


  const getSvgString = (iconName: string): string | null => {
    const svgStrings: { [key: string]: string } = {

      'default-star': `
        <svg width="24" height="24" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
          <path d="M12 2l3.09 6.26L22 9.27l-5 4.87 1.18 6.88L12 17.77l-6.18 3.25L7 14.14 2 9.27l6.91-1.01L12 2z" fill="${color || '#4A90E2'}"/>
        </svg>
      `,
    };

    return svgStrings[iconName] || null;
  };

  const iconPath = getIconPath(name);
  const svgString = getSvgString(name);


  if (svgString) {
    return (
      <View style={[styles.container, style]}>
        <SvgXml xml={svgString} width={size} height={size} />
      </View>
    );
  }


  if (iconPath) {
    const isImage = type === 'png' || (type === 'auto' && typeof iconPath === 'number');
    
    if (isImage) {

      return (
        <View style={[styles.container, style]}>
          <Image
            source={iconPath}
            style={[
              { width: size, height: size },
              color && { tintColor: color }
            ]}
            resizeMode="contain"
            fadeDuration={0}
          />
        </View>
      );
    } else {

      return (
        <View style={[styles.container, style]}>
          <SvgXml xml={iconPath} width={size} height={size} />
        </View>
      );
    }
  }

  // Fallback: return a placeholder
  return (
    <View style={[styles.container, styles.placeholder, { width: size, height: size }, style]}>
      <View style={[styles.placeholderInner, { backgroundColor: color || '#CCCCCC' }]} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  placeholder: {
    backgroundColor: '#F0F0F0',
    borderRadius: 4,
    alignItems: 'center',
    justifyContent: 'center',
  },
  placeholderInner: {
    width: '60%',
    height: '60%',
    borderRadius: 2,
  },
});
