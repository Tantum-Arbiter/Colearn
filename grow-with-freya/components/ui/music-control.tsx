import React from 'react';
import { View, TouchableOpacity, StyleSheet } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { useGlobalSound } from '@/contexts/global-sound-context';

interface MusicControlProps {
  size?: number;
  color?: string;
  style?: any;
}

export const MusicControl: React.FC<MusicControlProps> = ({
  size = 32,
  color = '#4A90E2',
  style
}) => {
  const { isMuted, volume, isLoaded, toggleMute } = useGlobalSound();

  const handlePress = () => {
    toggleMute();
  };

  return (
    <View style={[styles.container, style]}>
      <TouchableOpacity
        style={styles.iconContainer}
        onPress={handlePress}
        activeOpacity={0.7}
        accessibilityLabel={isMuted ? 'Unmute background music' : 'Mute background music'}
        accessibilityRole="button"
        testID="music-control-button"
      >
        <View style={[styles.iconBackground, { backgroundColor: `${color}20` }]}>
          <Ionicons
            name={isMuted ? 'volume-mute' : 'volume-high'}
            size={size}
            color={color}
            testID={`music-icon-${isMuted ? 'muted' : 'playing'}`}
          />
        </View>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  iconContainer: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  iconBackground: {
    width: 48,
    height: 48,
    borderRadius: 24,
    alignItems: 'center',
    justifyContent: 'center',
  },
});
