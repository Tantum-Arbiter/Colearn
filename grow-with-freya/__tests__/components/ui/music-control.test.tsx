import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { MusicControl } from '@/components/ui/music-control';

// Mock the background music hook
const mockToggle = jest.fn();
const mockUseBackgroundMusic = {
  isPlaying: false,
  isLoaded: true,
  volume: 0.3,
  play: jest.fn(),
  pause: jest.fn(),
  stop: jest.fn(),
  setVolume: jest.fn(),
  fadeIn: jest.fn(),
  fadeOut: jest.fn(),
  toggle: mockToggle,
};

jest.mock('@/hooks/use-background-music', () => ({
  useBackgroundMusic: () => mockUseBackgroundMusic,
}));

// Mock Ionicons
jest.mock('@expo/vector-icons', () => ({
  Ionicons: ({ name, ...props }: any) => {
    const { Text } = require('react-native');
    return <Text {...props}>{name}</Text>;
  },
}));

describe('MusicControl', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render when music is loaded', () => {
    const { getByLabelText } = render(<MusicControl />);
    expect(getByLabelText('Play background music')).toBeTruthy();
  });

  it('should not render when music is not loaded', () => {
    mockUseBackgroundMusic.isLoaded = false;
    const { queryByLabelText } = render(<MusicControl />);
    expect(queryByLabelText('Play background music')).toBeNull();

    // Reset for other tests
    mockUseBackgroundMusic.isLoaded = true;
  });

  it('should show mute icon when music is not playing', () => {
    mockUseBackgroundMusic.isPlaying = false;
    const rendered = render(<MusicControl />);
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it('should show volume icon when music is playing', () => {
    mockUseBackgroundMusic.isPlaying = true;
    const rendered = render(<MusicControl />);
    expect(rendered.toJSON()).toMatchSnapshot();

    // Reset for other tests
    mockUseBackgroundMusic.isPlaying = false;
  });

  it('should call toggle when pressed', () => {
    const { getByLabelText } = render(<MusicControl />);
    const button = getByLabelText(/background music/i);
    fireEvent.press(button);
    expect(mockToggle).toHaveBeenCalledTimes(1);
  });

  it('should have correct accessibility labels', () => {
    mockUseBackgroundMusic.isPlaying = false;
    const { getByLabelText } = render(<MusicControl />);
    expect(getByLabelText('Play background music')).toBeTruthy();

    mockUseBackgroundMusic.isPlaying = true;
    const { rerender, getByLabelText: getByLabelTextAfterRerender } = render(<MusicControl />);
    rerender(<MusicControl />);
    expect(getByLabelTextAfterRerender('Pause background music')).toBeTruthy();
  });

  it('should accept custom props', () => {
    mockUseBackgroundMusic.isPlaying = false;
    const rendered = render(
      <MusicControl size={48} color="#FF0000" />
    );
    expect(rendered.toJSON()).toMatchSnapshot();
    // Note: Testing custom props is limited with current mock setup
    // The size and color props are passed to the mock but not easily testable
  });
});
