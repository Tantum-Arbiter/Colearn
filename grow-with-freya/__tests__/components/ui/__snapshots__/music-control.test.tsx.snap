// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MusicControl should accept custom props 1`] = `
<button
  aria-label="Play background music"
  className="css-view-g5y9jx r-transitionProperty-1i6wzkk r-userSelect-lrvibr r-cursor-1loqt21 r-touchAction-1otgn73 r-alignItems-1awozwy r-justifyContent-1777fci"
  data-testid="music-control-button"
  dir={null}
  onClick={[Function]}
  onContextMenu={[Function]}
  onKeyDown={[Function]}
  ref={[Function]}
  role="button"
  style={
    {
      "transitionDuration": "0s",
    }
  }
  type="button"
>
  <div
    className="css-view-g5y9jx r-alignItems-1awozwy r-borderRadius-1fuqb1j r-boxShadow-1a0psd9 r-height-h3s6tt r-justifyContent-1777fci r-width-rwqe4o"
    dir={null}
    ref={[Function]}
    style={
      {
        "backgroundColor": "rgba(255,0,0,0.13)",
      }
    }
  >
    <div
      className="css-text-146c3p1"
      data-testid="music-icon-muted"
      dir="auto"
      ref={[Function]}
    >
      volume-mute
    </div>
  </div>
</button>
`;

exports[`MusicControl should show mute icon when music is not playing 1`] = `
<button
  aria-label="Play background music"
  className="css-view-g5y9jx r-transitionProperty-1i6wzkk r-userSelect-lrvibr r-cursor-1loqt21 r-touchAction-1otgn73 r-alignItems-1awozwy r-justifyContent-1777fci"
  data-testid="music-control-button"
  dir={null}
  onClick={[Function]}
  onContextMenu={[Function]}
  onKeyDown={[Function]}
  ref={[Function]}
  role="button"
  style={
    {
      "transitionDuration": "0s",
    }
  }
  type="button"
>
  <div
    className="css-view-g5y9jx r-alignItems-1awozwy r-borderRadius-1fuqb1j r-boxShadow-1a0psd9 r-height-h3s6tt r-justifyContent-1777fci r-width-rwqe4o"
    dir={null}
    ref={[Function]}
    style={
      {
        "backgroundColor": "rgba(74,144,226,0.13)",
      }
    }
  >
    <div
      className="css-text-146c3p1"
      data-testid="music-icon-muted"
      dir="auto"
      ref={[Function]}
    >
      volume-mute
    </div>
  </div>
</button>
`;

exports[`MusicControl should show volume icon when music is playing 1`] = `
<button
  aria-label="Pause background music"
  className="css-view-g5y9jx r-transitionProperty-1i6wzkk r-userSelect-lrvibr r-cursor-1loqt21 r-touchAction-1otgn73 r-alignItems-1awozwy r-justifyContent-1777fci"
  data-testid="music-control-button"
  dir={null}
  onClick={[Function]}
  onContextMenu={[Function]}
  onKeyDown={[Function]}
  ref={[Function]}
  role="button"
  style={
    {
      "transitionDuration": "0s",
    }
  }
  type="button"
>
  <div
    className="css-view-g5y9jx r-alignItems-1awozwy r-borderRadius-1fuqb1j r-boxShadow-1a0psd9 r-height-h3s6tt r-justifyContent-1777fci r-width-rwqe4o"
    dir={null}
    ref={[Function]}
    style={
      {
        "backgroundColor": "rgba(74,144,226,0.13)",
      }
    }
  >
    <div
      className="css-text-146c3p1"
      data-testid="music-icon-playing"
      dir="auto"
      ref={[Function]}
    >
      volume-high
    </div>
  </div>
</button>
`;
